library/core/src/slice/cmp.rs

```diff
@@ -55,12 +55,45 @@ impl<A, B> SlicePartialEq<B> for [A]
 where
     A: PartialEq<B>,
 {
+    #[inline]
     default fn equal(&self, other: &[B]) -> bool {
         if self.len() != other.len() {
             return false;
         }
 
-        self.iter().zip(other.iter()).all(|(x, y)| x == y)
+        // ZSTs have no identity and slices don't guarantee which addresses-to-ZSTs they produce
+        // so we only need to compare them once to determine the behavior of the PartialEq impl
```

Comment:
The `eq` could still have side-effects though, no?

We have <https://rust-lang.github.io/rfcs/1521-copy-clone-semantics.html> to let us skip side-effects in `clone` sometimes, but I don't think we have a justification to skip side-effects in `PartialEq`.